tests/TestCase/Controller/CookieComponentControllerTest.php

```diff
@@ -0,0 +1,124 @@
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP Project
+ * @since         3.1.6
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Test\TestCase\Controller;
+
+use Cake\Controller\Component;
+use Cake\Core\Configure;
+use Cake\Routing\DispatcherFactory;
+use Cake\Routing\Router;
+use Cake\TestSuite\IntegrationTestCase;
+use Cake\Utility\Security;
+
+/**
+ * CookieComponentControllerTest class
+ */
+class CookieComponentControllerTest extends IntegrationTestCase
+{
+    /**
+     * reset environment.
+     *
+     * @return void
+     */
+    public function setUp()
+    {
+        parent::setUp();
+        Configure::write('App.namespace', 'TestApp');
+
+        Security::salt('abcdabcdabcdabcdabcdabcdabcdabcdabcd');
+        Router::connect('/:controller/:action/*', [], ['routeClass' => 'InflectedRoute']);
+        DispatcherFactory::clear();
+        DispatcherFactory::add('Routing');
+        DispatcherFactory::add('ControllerFactory');
+    }
+
+    /**
+     * Can encrypt/decrypt the cookie value.
+     */
+    public function testCanEncryptAndDecryptWithAes()
+    {
+        $this->cookieEncrypted('NameOfCookie', 'Value of Cookie', 'aes');
+
+        $this->get('/cookie_component_test/view/');
+        $this->assertStringStartsWith('Q2FrZQ==.', $this->viewVariable('ValueFromRequest'), 'Encrypted');
+        $this->assertEquals('Value of Cookie', $this->viewVariable('ValueFromCookieComponent'), 'Decrypted');
+    }
+
+    /**
+     * Can encrypt/decrypt the cookie value by default.
+     */
+    public function testCanEncryptAndDecryptCookieValue()
+    {
+        $this->cookieEncrypted('NameOfCookie', 'Value of Cookie');
+
+        $this->get('/cookie_component_test/view/');
+        $this->assertStringStartsWith('Q2FrZQ==.', $this->viewVariable('ValueFromRequest'), 'Encrypted');
+        $this->assertEquals('Value of Cookie', $this->viewVariable('ValueFromCookieComponent'), 'Decrypted');
+    }
+
+    /**
+     * Can encrypt/decrypt even if the cookie value are array.
+     */
+    public function testCanEncryptAndDecryptEvenIfCookieValueIsArray()
+    {
+        $this->cookieEncrypted('NameOfCookie', ['Value1 of Cookie', 'Value2 of Cookie']);
+
+        $this->get('/cookie_component_test/view/');
+        $this->assertStringStartsWith('Q2FrZQ==.', $this->viewVariable('ValueFromRequest'), 'Encrypted');
+        $this->assertEquals(['Value1 of Cookie', 'Value2 of Cookie'], $this->viewVariable('ValueFromCookieComponent'),
+            'Decrypted');
+    }
+
+    /**
+     * @return array
+     */
+    public function encryptionKeyCases()
+    {
+        return [
+            ['foo!foo!foo!foo!foo!foo!foo!foo!'],
+            ['bar!bar!bar!bar!bar!bar!bar!bar!'],
+        ];
```

Comment:
I thought this is to avoid the hard coding salt. But now, I realize it certainly has no meaning. I will remove this dataProvider. Can I use `foo!foo!foo!foo!foo!foo!foo!foo!` as the salt on test case?